File: dlc-datastruct/queue-stack/queuq-bfs&dfs/752.打开转盘锁.go
/*
 * @lc app=leetcode.cn id=752 lang=golang
 *
 * [752] 打开转盘锁
 *
 * https://leetcode-cn.com/problems/open-the-lock/description/
 *
 * algorithms
 * Medium (48.02%)
 * Likes:    64
 * Dislikes: 0
 * Total Accepted:    6.3K
 * Total Submissions: 13.2K
 * Testcase Example:  '["0201","0101","0102","1212","2002"]\n"0202"'
 *
 * 你有一个带有四个圆形拨轮的转盘锁。每个拨轮都有10个数字： '0', '1', '2', '3', '4', '5', '6', '7', '8',
 * '9' 。每个拨轮可以自由旋转：例如把 '9' 变为  '0'，'0' 变为 '9' 。每次旋转都只能旋转一个拨轮的一位数字。
 *
 * 锁的初始数字为 '0000' ，一个代表四个拨轮的数字的字符串。
 *
 * 列表 deadends 包含了一组死亡数字，一旦拨轮的数字和列表里的任何一个元素相同，这个锁将会被永久锁定，无法再被旋转。
 *
 * 字符串 target 代表可以解锁的数字，你需要给出最小的旋转次数，如果无论如何不能解锁，返回 -1。
 *
 *
 *
 * 示例 1:
 *
 *
 * 输入：deadends = ["0201","0101","0102","1212","2002"], target = "0202"
 * 输出：6
 * 解释：
 * 可能的移动序列为 "0000" -> "1000" -> "1100" -> "1200" -> "1201" -> "1202" -> "0202"。
 * 注意 "0000" -> "0001" -> "0002" -> "0102" -> "0202" 这样的序列是不能解锁的，
 * 因为当拨动到 "0102" 时这个锁就会被锁定。
 *
 *
 * 示例 2:
 *
 *
 * 输入: deadends = ["8888"], target = "0009"
 * 输出：1
 * 解释：
 * 把最后一位反向旋转一次即可 "0000" -> "0009"。
 *
 *
 * 示例 3:
 *
 *
 * 输入: deadends = ["8887","8889","8878","8898","8788","8988","7888","9888"],
 * target = "8888"
 * 输出：-1
 * 解释：
 * 无法旋转到目标数字且不被锁定。
 *
 *
 * 示例 4:
 *
 *
 * 输入: deadends = ["0000"], target = "8888"
 * 输出：-1
 *
 *
 *
 *
 * 提示：
 *
 *
 * 死亡列表 deadends 的长度范围为 [1, 500]。
 * 目标数字 target 不会在 deadends 之中。
 * 每个 deadends 和 target 中的字符串的数字会在 10,000 个可能的情况 '0000' 到 '9999' 中产生。
 *
 *
 */

// @lc code=start
import "strconv"

func openLock(deadends []string, target string) int {
	//bfs
	h := make(map[string]bool)
	for _, ded := range deadends {
		h[ded] = true
	}

	if h[target] || h["0000"] {
		return -1
	}

	visit := make(map[string]bool)
	queue := []string{"0000"}
	step := 0
	for len(queue) > 0 {
		step++
		size := len(queue)
		for i := 0; i < size; i++ {
			cur := queue[0]
			queue = queue[1:]
			// 一共有四位数 idx表示第几位
			for idx := 0; idx < 4; idx++ {
				// 往上往下拨
				for j := -1; j <= 1; j += 2 {
					base, _ := strconv.Atoi(string(cur[idx]))
					base = (base + j + 10) % 10
					next := string(cur[:idx]) + strconv.Itoa(base) + string(cur[idx+1:])
					if next == target {
						return step
					}
					if h[next] || visit[next] {
						continue
					}
					queue = append(queue, next)
					visit[next] = true
				}
			}
		}
	}
	return -1
}

// @lc code=end

File: labuladong/开始之前/bfs/752.打开转盘锁.py
#
# @lc app=leetcode.cn id=752 lang=python3
#
# [752] 打开转盘锁
#

# @lc code=start
class Solution:
    def up(self, s, idx):
        """将指定位置的数向上波动一次"""
        t = int(s[idx])
        if t == 9:
            t = 0
        else:
            t += 1
        return s[:idx] + str(t) + s[idx + 1 :]

    def down(self, s, idx):
        """将指定位置的数向下波动一次"""
        t = int(s[idx])
        if t == 0:
            t = 9
        else:
            t -= 1
        return s[:idx] + str(t) + s[idx + 1 :]

    def openLock(self, deadends, target: str):
        """
        仔细想想，这就可以抽象成一幅图，每个节点有 8 个相邻的节点，又让你求最短距离
        """
        q = ["0000"]
        step = 0
        deadends = set(deadends)
        visited = set()

        while len(q) > 0:
            size = len(q)
            for i in range(0, size):
                cur = q.pop(0)
                if cur in deadends:
                    continue
                if cur == target:
                    return step

                # 开始拨动锁盘
                for i in range(0, len(cur)):
                    up = self.up(cur, i)
                    if up not in visited:
                        q.append(up)
                        visited.add(up)

                    down = self.down(cur, i)
                    if down not in visited:
                        q.append(down)
                        visited.add(down)
            step += 1

        return -1

    def test(self, deadends=["0201", "0101", "0102", "1212", "2002"], target="0202"):
        return self.openLock(deadends=deadends, target=target)


# @lc code=end

